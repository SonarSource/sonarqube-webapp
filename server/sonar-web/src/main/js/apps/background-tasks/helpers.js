/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const ONE_SECOND = 1000;
const ONE_MINUTE = 60 * ONE_SECOND;

function format (int, suffix) {
  return `${int}${suffix}`;
}

export function formatDuration (value) {
  if (!value) {
    return '';
  }
  if (value >= ONE_MINUTE) {
    const minutes = Math.round(value / ONE_MINUTE);
    return format(minutes, 'min');
  } else if (value >= ONE_SECOND) {
    const seconds = Math.round(value / ONE_SECOND);
    return format(seconds, 's');
  } else {
    return format(value, 'ms');
  }
}
