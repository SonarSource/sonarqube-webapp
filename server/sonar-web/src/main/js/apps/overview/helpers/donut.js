import d3 from 'd3';
import React from 'react';

let Sector = React.createClass({
  render() {
    let arc = d3.svg.arc()
        .outerRadius(this.props.radius)
        .innerRadius(this.props.radius - this.props.thickness);
    return <path d={arc(this.props.data)} style={{ fill: this.props.fill }}/>;
  }
});

export default React.createClass({
  getDefaultProps() {
    return {
      size: 30,
      thickness: 6
    };
  },

  render() {
    let radius = this.props.size / 2;
    let pie = d3.layout.pie()
        .sort(null)
        .value(d => d.value);
    let data = this.props.data;
    let sectors = pie(data).map((d, i) => {
      return <Sector key={i} data={d} fill={data[i].fill} radius={radius} thickness={this.props.thickness}/>;
    });
    return <svg width={this.props.size} height={this.props.size}>
      <g transform={`translate(${radius}, ${radius})`}>{sectors}</g>
    </svg>;
  }
});
