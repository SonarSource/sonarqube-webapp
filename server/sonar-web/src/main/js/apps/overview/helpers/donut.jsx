import React from 'react';

const Sector = React.createClass({
  render() {
    const arc = d3.svg.arc()
        .outerRadius(this.props.radius)
        .innerRadius(this.props.radius - this.props.thickness);
    return <path d={arc(this.props.data)} style={{ fill: this.props.fill }}></path>;
  }
});

export default React.createClass({
  getDefaultProps() {
    return {
      size: 30,
      thickness: 6
    };
  },

  render() {
    const radius = this.props.size / 2;
    const pie = d3.layout.pie().sort(null)
        .value(d => {
          return d.value
        });
    const data = this.props.data;
    const sectors = pie(data).map((d, i) => {
      return <Sector
          key={i}
          data={d}
          fill={data[i].fill}
          radius={radius}
          thickness={this.props.thickness}/>;
    });
    return (
        <svg width={this.props.size} height={this.props.size}>
          <g transform={`translate(${radius}, ${radius})`}>{sectors}</g>
        </svg>
    );
  }
});
