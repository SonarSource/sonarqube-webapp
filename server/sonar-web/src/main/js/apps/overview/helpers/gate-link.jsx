import React from 'react';

export default React.createClass({
  render() {
    const url = `${baseUrl}/quality_gates/show/${this.props.gate}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
