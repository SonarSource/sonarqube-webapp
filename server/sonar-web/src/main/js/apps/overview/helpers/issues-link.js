import React from 'react';

export default React.createClass({
  render() {
    let params = Object.keys(this.props.params).map((key) => {
          return `${key}=${encodeURIComponent(this.props.params[key])}`;
        }).join('|'),
        url = `${baseUrl}/component_issues/index?id=${encodeURIComponent(this.props.component)}#${params}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
