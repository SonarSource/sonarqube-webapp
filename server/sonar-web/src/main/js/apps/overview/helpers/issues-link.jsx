import React from 'react';

export default React.createClass({
  render: function () {
    var params = Object.keys(this.props.params).map((key) => {
      return `${key}=${encodeURIComponent(this.props.params[key])}`;
    }).join('|');
    var url = `${baseUrl}/component_issues/index?id=${encodeURIComponent(this.props.component)}#${params}`;
    return <a href={url}>{this.props.children}</a>;
  }
});
