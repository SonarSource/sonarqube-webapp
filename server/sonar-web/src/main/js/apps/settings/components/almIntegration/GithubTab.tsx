/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { createGithubConfiguration, updateGithubConfiguration } from '../../../../api/alm-settings';
import {
  AlmKeys,
  AlmSettingsBindingStatus,
  GithubBindingDefinition
} from '../../../../types/alm-settings';
import { ALM_INTEGRATION } from '../AdditionalCategoryKeys';
import CategoryDefinitionsList from '../CategoryDefinitionsList';
import AlmTab from './AlmTab';
import GithubForm from './GithubForm';

export interface GithubTabProps {
  branchesEnabled: boolean;
  component?: T.Component;
  definitions: GithubBindingDefinition[];
  definitionStatus: T.Dict<AlmSettingsBindingStatus>;
  loadingAlmDefinitions: boolean;
  loadingProjectCount: boolean;
  multipleAlmEnabled: boolean;
  onCheck: (definitionKey: string) => void;
  onDelete: (definitionKey: string) => void;
  onUpdateDefinitions: () => void;
}

export default function GithubTab(props: GithubTabProps) {
  const {
    branchesEnabled,
    component,
    multipleAlmEnabled,
    definitions,
    definitionStatus,
    loadingAlmDefinitions,
    loadingProjectCount
  } = props;

  return (
    <div className="bordered">
      {branchesEnabled && (
        <>
          <AlmTab
            alm={AlmKeys.GitHub}
            createConfiguration={createGithubConfiguration}
            defaultBinding={{
              key: '',
              appId: '',
              clientId: '',
              clientSecret: '',
              url: '',
              privateKey: ''
            }}
            definitions={definitions}
            definitionStatus={definitionStatus}
            form={childProps => <GithubForm {...childProps} />}
            loadingAlmDefinitions={loadingAlmDefinitions}
            loadingProjectCount={loadingProjectCount}
            multipleAlmEnabled={multipleAlmEnabled}
            onCheck={props.onCheck}
            onDelete={props.onDelete}
            onUpdateDefinitions={props.onUpdateDefinitions}
            updateConfiguration={updateGithubConfiguration}
          />

          <div className="huge-spacer-top huge-spacer-bottom bordered-top" />
        </>
      )}

      <div className="big-padded">
        <CategoryDefinitionsList
          category={ALM_INTEGRATION}
          component={component}
          subCategory={AlmKeys.GitHub}
        />
      </div>
    </div>
  );
}
