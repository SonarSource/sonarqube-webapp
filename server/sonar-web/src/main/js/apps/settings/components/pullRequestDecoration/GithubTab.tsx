/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { createGithubConfiguration, updateGithubConfiguration } from '../../../../api/almSettings';
import { GithubBindingDefinition } from '../../../../types/alm-settings';
import GithubTabRenderer from './GithubTabRenderer';

interface Props {
  definitions: GithubBindingDefinition[];
  loading: boolean;
  onDelete: (definitionKey: string) => void;
  onUpdateDefinitions: () => void;
}

interface State {
  editedDefinition?: GithubBindingDefinition;
  projectCount?: number;
}

export default class GithubTab extends React.PureComponent<Props, State> {
  mounted = false;
  state: State = {};

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleCancel = () => {
    this.setState({
      editedDefinition: undefined
    });
  };

  handleCreate = () => {
    this.setState({ editedDefinition: { key: '', appId: '', url: '', privateKey: '' } });
  };

  handleEdit = (definitionKey: string) => {
    const editedDefinition = this.props.definitions.find(d => d.key === definitionKey);
    this.setState({ editedDefinition });
  };

  handleSubmit = (config: GithubBindingDefinition, originalKey: string) => {
    const call = originalKey
      ? updateGithubConfiguration({ newKey: config.key, ...config, key: originalKey })
      : createGithubConfiguration(config);
    return call
      .then(() => {
        if (this.mounted) {
          this.setState({ editedDefinition: undefined });
        }
      })
      .then(this.props.onUpdateDefinitions);
  };

  render() {
    const { definitions, loading } = this.props;
    const { editedDefinition } = this.state;
    return (
      <GithubTabRenderer
        definitions={definitions}
        editedDefinition={editedDefinition}
        loading={loading}
        onCancel={this.handleCancel}
        onCreate={this.handleCreate}
        onDelete={this.props.onDelete}
        onEdit={this.handleEdit}
        onSubmit={this.handleSubmit}
      />
    );
  }
}
