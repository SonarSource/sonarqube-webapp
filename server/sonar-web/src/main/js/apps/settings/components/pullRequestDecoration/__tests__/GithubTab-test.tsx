/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import {
  createGithubConfiguration,
  deleteConfiguration,
  updateGithubConfiguration
} from '../../../../../api/almSettings';
import { mockGithubDefinition } from '../../../../../helpers/testMocks';
import GithubTab from '../GithubTab';

jest.mock('../../../../../api/almSettings', () => ({
  countBindedProjects: jest.fn().mockResolvedValue(2),
  createGithubConfiguration: jest.fn().mockResolvedValue({}),
  deleteConfiguration: jest.fn().mockResolvedValue({}),
  updateGithubConfiguration: jest.fn().mockResolvedValue({})
}));

beforeEach(() => {
  jest.clearAllMocks();
});

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('should handle cancel', async () => {
  const wrapper = shallowRender();

  wrapper.setState({
    definitionKeyForDeletion: '12321',
    definitionInEdition: mockGithubDefinition()
  });

  wrapper.instance().handleCancel();

  await waitAndUpdate(wrapper);

  expect(wrapper.state().definitionKeyForDeletion).toBeUndefined();
  expect(wrapper.state().definitionInEdition).toBeUndefined();
});

it('should delete config', async () => {
  const onUpdateDefinitions = jest.fn();
  const wrapper = shallowRender({ onUpdateDefinitions });
  wrapper.setState({ definitionKeyForDeletion: '123' });

  await wrapper
    .instance()
    .deleteConfiguration('123')
    .then(() => {
      expect(deleteConfiguration).toBeCalledWith('123');
      expect(onUpdateDefinitions).toBeCalled();
      expect(wrapper.state().definitionKeyForDeletion).toBeUndefined();
    });
});

it('should create config', async () => {
  const onUpdateDefinitions = jest.fn();
  const config = {
    key: 'new conf',
    url: 'ewrqewr',
    appId: '3742985',
    privateKey: 'rt7r78ew6t87ret'
  };
  const wrapper = shallowRender({ onUpdateDefinitions });
  wrapper.setState({ definitionInEdition: config });

  await wrapper
    .instance()
    .handleSubmit(config, '')
    .then(() => {
      expect(createGithubConfiguration).toBeCalledWith(config);
      expect(onUpdateDefinitions).toBeCalled();
      expect(wrapper.state().definitionInEdition).toBeUndefined();
    });
});

it('should update config', async () => {
  const onUpdateDefinitions = jest.fn();
  const config = {
    key: 'new conf',
    url: 'ewrqewr',
    appId: '3742985',
    privateKey: 'rt7r78ew6t87ret'
  };
  const wrapper = shallowRender({ onUpdateDefinitions });
  wrapper.setState({ definitionInEdition: config });

  await wrapper
    .instance()
    .handleSubmit(config, 'originalKey')
    .then(() => {
      expect(updateGithubConfiguration).toBeCalledWith({
        newKey: 'new conf',
        ...config,
        key: 'originalKey'
      });
      expect(onUpdateDefinitions).toBeCalled();
      expect(wrapper.state().definitionInEdition).toBeUndefined();
    });
});

function shallowRender(props: Partial<GithubTab['props']> = {}) {
  return shallow<GithubTab>(
    <GithubTab definitions={[]} onUpdateDefinitions={jest.fn()} {...props} />
  );
}
