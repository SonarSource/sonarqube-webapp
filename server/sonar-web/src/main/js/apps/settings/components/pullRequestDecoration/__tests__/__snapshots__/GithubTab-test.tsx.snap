// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<TabRenderer
  alm="github"
  definitions={Array []}
  onCancel={[Function]}
  onConfirmDelete={[Function]}
  onCreate={[Function]}
  onDelete={[Function]}
  onEdit={[Function]}
  onSubmit={[Function]}
/>
`;
